### PR TITLE
class library: Fix BinaryOpUGen optimization for a === b cases

### PR DESCRIPTION
**Pull request by smrg-lm**
## Purpose and Motivation

This is an attempt to fix issue #5405

## Types of changes

- Bug fix

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] This PR is ready for review

### CODE REVIEW

**Review by jamshark70 — CHANGES_REQUESTED** [review]
At a quick glance -- if a == b, then the entire optimization is invalid and nothing should be done. That is, the optimization is to remove upstream unit(s). There's no particular benefit to modifying the current operator being optimized *without* removing at least one input.

So I would suggest, e.g. `optimizeAddNeg` (just because that is the first in the diff):

```
	optimizeAddNeg {
		var a, b, replacement;
		#a, b = inputs;

		if(a === b) { ^nil };  // <<-- I think this is enough
		
		if (b.isKindOf(UnaryOpUGen) and: { b.operator == 'neg' and: { b.descendants.size == 1 } }) {
			... snip
		}

		... snip

		^nil
	}
```

**Comment by jamshark70** [comment]
Adding to yesterday's comment -- on second thought, modifying one unit without removing the other would be worse for performance. With the possible exception of `optimizeSub`, the optimizations take two operator units and replace them with one, more complex operation -- so the entire optimization is the removal of one of the source operators.

In `optimizeToMulAdd`:

- Given unit 1 = a*b and unit 2 = u1 + c (and u1 is not used anywhere else), a correct optimization is to delete u1 and replace u2 with MulAdd(a, b, c). `*` and `+` are still done only once, but in one function (saving function call overhead).

- Given unit 1 = a*b and unit 2 = u1 + u1, suppose you do not delete u1, but only replace u2 with MulAdd(a, a, u1). Now the multiplication is being done twice: once in u1, and again in MulAdd, but the function call overhead is unchanged! That's an anti-optimization.

So it's not only that "there's no particular benefit" -- it would actually make matters worse.

**Comment by smrg-lm** [comment]
> In `optimizeToMulAdd`:
> 
>     * Given unit 1 = a*b and unit 2 = u1 + c (and u1 is not used anywhere else), a correct optimization is to delete u1 and replace u2 with MulAdd(a, b, c). `*` and `+` are still done only once, but in one function (saving function call overhead).
> 
>     * Given unit 1 = a*b and unit 2 = u1 + u1, suppose you do not delete u1, but only replace u2 with MulAdd(a, a, u1). Now the multiplication is being done twice: once in u1, and again in MulAdd, but the function call overhead is unchanged! That's an anti-optimization.

This cases are not valid because u1 must be u2. However I thought about what you mean and posted resulting optimizations in #5405

For example:

```supercollider
// optimizeToSum4
x = SinOsc.ar([100, 200, 300]).sum();
x + x

['0_SinOsc', 'audio', [100, 0.0]]
['1_SinOsc', 'audio', [200, 0.0]]
['2_SinOsc', 'audio', [300, 0.0]]
['3_Sum3', 'audio', ['0_SinOsc', '1_SinOsc', '2_SinOsc']]
['4_Sum4', 'audio', ['0_SinOsc', '1_SinOsc', '2_SinOsc', '3_Sum3']]
['5_Out', 'audio', [0, '4_Sum4']]
```
Results in a Sum4 and a Sum3 instead of the sum of two Sum3, I'm don't know what would be more optimal, I would guess the Sum4 because there is one less opcode/ugen.

```supercollider
// optimizeToMulAdd
x = SinOsc.ar(100) * 0.1;
x + x

['0_SinOsc', 'audio', [100, 0.0]]
['1_*', 'audio', ['0_SinOsc', 0.1]]
['2_MulAdd', 'audio', ['0_SinOsc', 0.1, '1_*']]
['3_Out', 'audio', [0, '2_MulAdd']]
```
Compared to 2_+ ... [1_*, 1_*] is less clear if there is any benefit because this case multiplies twice 0_SinOsc and 1 in 1_* and MulAdd. I think is not good.

```supercollider
// optimizeAddNeg
x = SinOsc.ar(100).neg;
x + x

['0_SinOsc', 'audio', [100, 0.0]]
['1_neg', 'audio', ['0_SinOsc']]
['2_-', 'audio', ['1_neg', '0_SinOsc']]
['3_Out', 'audio', [0, '2_-']]
```
This case I think it is 2_+ ...[1_neg, 1_neg] instead, same as the case before.

So, yes, the gain seems to be nothing but doesn't seem to be clearly worst. It is even more simple and clean to just ignore the optimization and just return nil for each edge case.

However, notice that there are other possible optimizations for cases when u1 is u2 but require replacement and adding new ugens which is not allowed while rewriting.

I would got for just returning nil in the edge cases and to ignore the optimization in each place the identity check is done is this patch if that is correct.

**Comment by jamshark70** [comment]
> This cases are not valid because u1 must be u2

Well... read it again. I'm quite sure my logic is correct.

The first example is a case where optimization is valid. Optimization is valid when u1 !== u2. So if I want to discuss a valid optimization case, then I *must* present a case where u1 !== u2, in order to contrast it against the invalid optimization.

Then the second example is an invalid optimization, and as you'll note in that example, I specify that u1 === u2 (`u1 + u1`).

> ['3_Sum3', 'audio', ['0_SinOsc', '1_SinOsc', '2_SinOsc']]
> ['4_Sum4', 'audio', ['0_SinOsc', '1_SinOsc', '2_SinOsc', '3_Sum3']]

This is definitely suboptimal because it is repeating the a+b+c in multiple units.

> Compared to 2_+ ... [1_, 1_] is less clear if there is any benefit because this case multiplies twice 0_SinOsc and 1 in 1_* and MulAdd. I think is not good.

Exactly -- it's not good. That's precisely what I was saying before.

> I would got for just returning nil in the edge cases and to ignore the optimization in each place the identity check is done is this patch if that is correct.

Agreed, yes, completely skip the optimization (no substitution).

**Comment by smrg-lm** [comment]
> > This cases are not valid because u1 must be u2
> 
> Well... read it again. I'm quite sure my logic is correct.

My apologies, I don't know what I read. It does replace operations between ugens by operation inside ugens and repeats things already done.

I will change it to skip optimizations.

File: SCClassLibrary/Common/Audio/BasicOpsUGen.sc [hunk 1]
```diff
@@ -171,8 +171,9 @@ BinaryOpUGen : BasicOpUGen {
 
 		if (b.isKindOf(UnaryOpUGen) and: { b.operator == 'neg' and: { b.descendants.size == 1 } }) {
 			// a + b.neg -> a - b
+			if(b === a) { ^nil };  // non optimizable edge case.
 			buildSynthDef.removeUGen(b);
-			replacement = a - b.inputs[0];
+			replacement = BinaryOpUGen('-', a, b.inputs[0]);
```
Comment (jamshark70) [inline SCClassLibrary/Common/Audio/BasicOpsUGen.sc:176]: While this change isn't harmful, it doesn't really improve anything either. I would leave this, and the next occurrence, alone.

File: SCClassLibrary/Common/Audio/BasicOpsUGen.sc [hunk 2]
```diff
@@ -202,13 +203,15 @@ BinaryOpUGen : BasicOpUGen {
 			and: { a.descendants.size == 1 }})
 		{
 			if (MulAdd.canBeMulAdd(a.inputs[0], a.inputs[1], b)) {
+				if(a === b) { ^nil };
```
Comment (jamshark70) [inline SCClassLibrary/Common/Audio/BasicOpsUGen.sc:206]: Come to think of it, all of these tests can be the first line in their respective methods. If a === b, there is no need to do any more complex checks -- check the simple thing first and bail out if possible.

```
	optimizeToMulAdd {
		var a, b, replacement;
		#a, b = inputs;

		if(a === b) { ^nil };
		if (a.isKindOf(BinaryOpUGen) and: { a.operator == '*'
... snip
```

**Review by jamshark70 — CHANGES_REQUESTED** [review]
In general I agree -- just some details.

`optimizeSub` is interesting case. If a === b, you would have the form `var x = someUGen.neg; x - x` but of course this can only be 0. (OK, well, `inf - inf == nan` but...) I thought it might even be possible to drop the entire chain in this case and replace with a `DC.perform(a.methodSelectorForRate, 0)` but I couldn't get it to play nicely with dead code elimination.

But even without the `if(a === b)` check, strangely, optimizeSub produces a valid result!

```
SynthDef(\x, {
	var x = SinOsc.ar(440).neg;
	Out.ar(0, x - x);
}).dumpUGens;

[ 0_SinOsc, audio, [ 440, 0.0 ] ]
[ 1_-, audio, [ 0_SinOsc, 0_SinOsc ] ]
[ 2_Out, audio, [ 0, 1_- ] ]

var x = SinOsc.ar;
var y = x.neg;

y - y

// remove 'y' (the 'neg' operator)
// and create: a + b.inputs[0]
y + y.inputs[0]
== y + x

// and this triggers an optimizeAddNeg!
--> x.neg + x
--> x - x
```

So you do actually end up with an optimization (`neg` is removed) and the result will be equivalent to the original (-x - -x == x - x) :exploding_head: 

So AFAICS, it's OK to leave optimizeSub as is.

File: SCClassLibrary/Common/Audio/BasicOpsUGen.sc (hunk 2 above)
Comment (smrg-lm) [reply SCClassLibrary/Common/Audio/BasicOpsUGen.sc:206]: Yes, now that I see it again I think is logically the same for all cases to put it before the other checks, I was following optimizeToSum3 but it makes not much sense for these other methods.

File: SCClassLibrary/Common/Audio/BasicOpsUGen.sc (hunk 1 above)
Comment (smrg-lm) [reply SCClassLibrary/Common/Audio/BasicOpsUGen.sc:176]: I thought it as proposal for review. It makes explicit use of the replacement object as done below by other method which also makes apparently dissimilar code (and the op gets hide to the view too). Can I keep it?

**Comment by smrg-lm** [comment]
> I thought it might even be possible to drop the entire chain in this case and replace with a `DC.perform(a.methodSelectorForRate, 0)` but I couldn't get it to play nicely with dead code elimination.

If you are using addUGen it is forbidden "// we don't add ugens, if a rewrite operation is in progress" and will ignore the addition to the graph. Later I thought that this code was done with a more narrow optimization scope and works fine within its limits, to add different cases could make the code more cryptic and other cases should be considered.
 
> But even without the `if(a === b)` check, strangely, optimizeSub produces a valid result!

IIRC that is because it always produces a case optimized by optimizeAddNeg as you noted. So I leaved as is, but now, in a third thought, it would be more correct to apply the check here too... I don't know, you can decide for me.

**Comment by jamshark70** [comment]
> joshpar assigned jamshark70 4 hours ago

This one will take some time to sift through. I might not be able to get to it quickly.

**Comment by joshpar** [comment]
Thanks James! I’ll pick up on it

/*
Josh Parmenter
www.realizedsound.net/josh
*/

**Comment by joshpar** [comment]
@smrg-lm - what are your thoughts on this now? Following through the convo and the logic (on the surface) this looks ok to me, but since the convo kind of dropped off I wanted to see if you felt like this is still good to go. If you say yes, I'll run a couple tests on my side and if that looks good I'll approve. But please let me know if any of your thinking has changed or if there is an edge case that you think is possibly a problem that we need to watch for. Thanks!

**Comment by smrg-lm** [comment]
I don't remember the state o this pr I'll take a look today if my back allow.

**Comment by smrg-lm** [comment]
@joshpar It should be ready to merge, the diff of the latest commits to the previous head is attached, it only adds checks and tests for the edge case that was causing the problem (without any refactoring), maybe you can squash them into one.

[commit.txt](https://github.com/supercollider/supercollider/files/8142418/commit.txt)

**Comment by joshpar** [comment]
@jamshark70 - how is your comfort level with this now? - we'd like to pull this in for 3.13.0 and are planning an RC for that after our meeting on Aug, 27, 2022. Anything else to add? Or can we close this?

**Comment by jamshark70** [comment]
I think this is fine. I'd say, merge it.

Removing improper optimizations is good. 

The only bad things I can think of here would be a/ perhaps it doesn't fix every related bug, or b/ I suppose there's a nonzero chance that this PR could remove proper optimizations, but it seems extremely unlikely.

Either of these possibilities would stem from the fact that UGen descendents are tracked currently per downstream UGen, instead of per downstream input. That would be a bigger change, and likely to have a performance impact.

**Comment by dyfer** [comment]
> I think this is fine. I'd say, merge it.

Would you approve it then? PR says that you requested changes :)

**Review by jamshark70 — APPROVED** [review]
